File: dataStructure/Java/src/main/java/sort/selectionSort/InsertionSort.java
package sort.selectionSort;

import sort.common.SortFunction;

/**
 * Created by lzhao on 7/20/17.
 */

public class InsertionSort implements SortFunction {
    public int[] sort(int[] array) {

        for (int i = 0; i < array.length; i++) {
            int pos = 0;
            int current = array[i];
            for (int j = i ; j >= 1 ; j--) {
                if (array[j - 1] > current) {
                    array[j] = array[j - 1];
                } else {
                    pos = j;
                    break;
                }
            }
            array[pos] = current;
        }

        return array;
    }
}

File: dataStructure/typescript/ts/src/sort/insertionSort.ts
export const insertionSort = <T>(arr: T[]): T[] => {
  const arrCopy = arr;

  for (let i = 1; i < arrCopy.length; i++) {
    const current = arrCopy[i];
    let pos = 0;
    for (let j = i; j > 0; j--) {
      if (current < arrCopy[j - 1]) {
        arrCopy[j] = arrCopy[j - 1];
      } else {
        pos = j;
        break;
      }
    }
    arrCopy[pos] = current;
  }

  return arrCopy;
};
